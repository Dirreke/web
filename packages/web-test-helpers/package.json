{
  "name": "web-test-helpers",
  "private": true,
  "peerDependencies": {
      "@vue/composition-api": "1.7.1",
      "vue-gettext": "^2.1.5",
      "vue-router": "3.6.5",
      "vuex": "^3.6.2"
  }
}
