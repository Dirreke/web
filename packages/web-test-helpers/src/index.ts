export * from './mocks/defaultComponentMocks'
export * from './mocks/defaultStubs'
export * from './mocks/axios'

export * from './defaultPlugins'
export * from './helpers'
