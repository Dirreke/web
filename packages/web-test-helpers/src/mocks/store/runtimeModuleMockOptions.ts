export const runtimeModuleMockOptions = {
  runtime: {
    namespaced: true,
    modules: {
      spaces: {
        namespaced: true,
        getters: {
          spaces: jest.fn(() => [])
        },
        mutations: {
          UPDATE_SPACE_FIELD: jest.fn()
        },
        actions: {
          loadSpaceMembers: jest.fn(),
          reloadProjectSpaces: jest.fn()
        }
      }
    }
  }
}
