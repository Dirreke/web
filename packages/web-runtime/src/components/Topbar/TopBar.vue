<template>
  <header
    id="oc-topbar"
    class="oc-flex oc-flex-middle oc-flex-between oc-pr-m oc-pl-s"
    :aria-label="$gettext('Top bar')"
  >
    <div class="oc-topbar-left oc-flex oc-flex-middle oc-flex-between">
      <applications-menu v-if="appMenuItems.length" :applications-list="appMenuItems" />
    </div>
    <div class="portal-wrapper">
      <portal-target name="app.runtime.header" multiple></portal-target>
    </div>
    <div class="oc-topbar-right oc-flex oc-flex-middle oc-flex-between">
      <portal-target name="app.runtime.header.right" multiple />
    </div>
    <portal to="app.runtime.header.right" :order="50">
        <oc-expanding-dropdown>
        <template #toggle>
          <oc-icon name="grid" size="large" class="oc-flex" />
        </template>
        <template #body>
          <oc-list class="applications-list">
            <div style="width: 200px; height: 150px">123</div>
          </oc-list>
        </template>
      </oc-expanding-dropdown>
      <theme-switcher v-if="darkThemeAvailable" />
      <feedback-link v-if="isFeedbackLinkEnabled" v-bind="feedbackLinkOptions" />
    </portal>
    <portal to="app.runtime.header.right" :order="100">
      <notifications v-if="isNotificationBellEnabled" />
      <user-menu v-if="isUserMenuEnabled" :applications-list="userMenuItems" />
    </portal>
  </header>
</template>

<script lang="ts">
import { mapGetters } from 'vuex'
import NavigationMixin from '../../mixins/navigationMixin'

import ApplicationsMenu from './ApplicationsMenu.vue'
import UserMenu from './UserMenu.vue'
import Notifications from './Notifications.vue'
import FeedbackLink from './FeedbackLink.vue'
import ThemeSwitcher from './ThemeSwitcher.vue'
import { useCapabilityNotifications, useStore, useUserContext } from 'web-pkg/src'
import { computed, unref } from 'vue'
import OcExpandingDropdown from '../../../../design-system/src/components/OcExpandingDropdown/OcExpandingDropdown.vue'

export default {
  components: {
    ApplicationsMenu,
    FeedbackLink,
    Notifications,
    ThemeSwitcher,
    UserMenu,
    OcExpandingDropdown
  },
  mixins: [NavigationMixin],
  props: {
    applicationsList: {
      type: Array,
      required: false,
      default: () => []
    }
  },
  setup() {
    const store = useStore()
    const notificationsSupport = useCapabilityNotifications()
    const isUserContext = useUserContext({ store })

    const isNotificationBellEnabled = computed(() => {
      return unref(isUserContext) && unref(notificationsSupport).includes('list')
    })

    return {
      isNotificationBellEnabled
    }
  },
  computed: {
    ...mapGetters(['configuration', 'user']),

    activeRoutePath() {
      return this.$router.resolve(this.$route).path
    },

    appMenuItems() {
      return this.navigation_getMenuItems([null, 'apps', 'appSwitcher'], this.activeRoutePath)
    },
    userMenuItems() {
      return this.navigation_getMenuItems(['user'], this.activeRoutePath)
    },

    darkThemeAvailable() {
      return this.configuration.themes.default && this.configuration.themes['default-dark']
    },

    sidebarLogoAlt() {
      return this.$gettext('Navigate to personal files page')
    },

    logoImage() {
      return this.configuration.currentTheme.logo.topbar
    },

    isFeedbackLinkEnabled() {
      return !this.configuration?.options?.disableFeedbackLink
    },

    feedbackLinkOptions() {
      const feedback = this.configuration?.options?.feedbackLink
      if (!this.isFeedbackLinkEnabled || !feedback) {
        return {}
      }

      return {
        ...(feedback.href && { href: feedback.href }),
        ...(feedback.ariaLabel && { ariaLabel: feedback.ariaLabel }),
        ...(feedback.description && { description: feedback.description })
      }
    },

    isUserMenuEnabled() {
      return this.user?.id
    }
  }
}
</script>

<style lang="scss">
#oc-topbar {
  height: 52px;
  position: sticky;
  z-index: 5;

  .portal-wrapper {
    @media (max-width: 639px) {
      margin-left: auto;
    }
  }

  img {
    max-height: 38px;
    image-rendering: auto;
    image-rendering: crisp-edges;
    image-rendering: pixelated;
    image-rendering: -webkit-optimize-contrast;
    user-select: none;
  }

  .oc-topbar-left {
    width: 10rem;
    gap: 30px;
  }
  .oc-topbar-right {
    gap: 20px;
  }
  @media only screen and (max-width: 960px) {
    .oc-topbar-left,
    .oc-topbar-right {
      gap: 0.5rem;
    }
  }
}
</style>
