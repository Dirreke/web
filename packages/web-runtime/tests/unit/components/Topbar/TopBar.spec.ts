import TopBar from 'web-runtime/src/components/Topbar/TopBar.vue'
import stubs from '../../../../../../tests/unit/stubs'
import { createStore, defaultComponentMocks, defaultPlugins, shallowMount } from 'web-test-helpers'
import { defaultStoreMockOptions } from 'web-test-helpers/src/mocks/store/defaultStoreMockOptions'

jest.spyOn((TopBar as any).mixins[0].methods, 'navigation_getMenuItems').mockImplementation(() => [
  {
    icon: 'folder',
    iconUrl: undefined,
    title: 'Files',
    path: '/files'
  },
  {
    icon: 'some-icon',
    iconUrl: undefined,
    title: 'External',
    url: 'http://some.org',
    target: '_blank'
  },
  {
    icon: 'application',
    iconUrl: undefined,
    path: '/settings',
    title: 'Settings'
  }
])

describe('Top Bar component', () => {
  it('Displays applications menu', () => {
    const { wrapper } = getWrapper()
    expect(wrapper.html().indexOf('applications-menu-stub')).toBeGreaterThan(-1)
    expect(wrapper).toMatchSnapshot()
  })
})

const getWrapper = () => {
  const mocks = { ...defaultComponentMocks() }
  const storeOptions = defaultStoreMockOptions
  storeOptions.getters.configuration.mockImplementation(() => ({
    options: { disableFeedbackLink: false },
    themes: {
      default: {},
      'default-dark': {}
    },
    currentTheme: {
      logo: {
        topbar: 'example-logo.svg'
      }
    }
  }))
  storeOptions.getters.user.mockImplementation(() => ({ id: 'einstein' }))
  const store = createStore(storeOptions)
  return {
    wrapper: shallowMount(TopBar, {
      props: {
        applicationsList: ['testApp']
      },
      global: {
        plugins: [...defaultPlugins(), store],
        stubs,
        mocks
      }
    })
  }
}
