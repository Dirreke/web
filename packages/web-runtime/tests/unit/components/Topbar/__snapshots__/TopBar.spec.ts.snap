// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Top Bar component Displays applications menu 1`] = `
<header aria-label="Top bar" class="oc-flex oc-flex-middle oc-flex-between oc-px-m" id="oc-topbar">
  <div class="oc-topbar-left oc-flex oc-flex-middle oc-flex-between">
    <applications-menu-stub applicationslist="[object Object],[object Object],[object Object]"></applications-menu-stub>
    <router-link-stub to="/">
      <oc-img-stub alt="Navigate to personal files page" class="oc-logo-image" loadingtype="eager" src="example-logo.svg"></oc-img-stub>
    </router-link-stub>
  </div>
  <div class="portal-wrapper">
    <portal-target-stub multiple name="app.runtime.header"></portal-target-stub>
  </div>
  <div class="oc-topbar-right oc-flex oc-flex-middle oc-flex-between">
    <theme-switcher-stub></theme-switcher-stub>
    <feedback-link-stub></feedback-link-stub>
    <!---->
    <user-menu-stub applicationslist="[object Object],[object Object],[object Object]"></user-menu-stub>
  </div>
</header>
`;
